src/Features/Core/Portable/InheritanceMargin/AbstractInheritanceMarginService_Helpers.cs

```diff
@@ -273,53 +271,42 @@ internal abstract partial class AbstractInheritanceMarginService
             ArrayBuilder<InheritanceMarginItem> builder,
             CancellationToken cancellationToken)
         {
-            // Get all base types.
-            var allBaseSymbols = BaseTypeFinder.FindBaseTypesAndInterfaces(memberSymbol);
-
-            // Filter out
-            // 1. System.Object. (otherwise margin would be shown for all classes)
-            // 2. System.ValueType. (otherwise margin would be shown for all structs)
-            // 3. System.Enum. (otherwise margin would be shown for all enum)
-            // 4. Error type.
-            // For example, if user has code like this,
-            // class Bar : ISomethingIsNotDone { }
-            // The interface has not been declared yet, so don't show this error type to user.
-            var baseSymbols = allBaseSymbols
-                .WhereAsArray(symbol => !symbol.IsErrorType() && symbol.SpecialType is not (SpecialType.System_Object or SpecialType.System_ValueType or SpecialType.System_Enum));
-
-            // Get all derived types
-            var allDerivedSymbols = await GetDerivedTypesAndImplementationsAsync(
-                solution,
+            var (baseTypeSymbolGroups, baseInterfaceSymbolGroups) = await BaseTypeSymbolsFinder.Instance.GetBaseTypeAndBaseInterfaceSymbolGroupsAsync(
+                    memberSymbol, solution, cancellationToken).ConfigureAwait(false);
+
+            var derivedTypeSymbolGroups = await DerivedTypeSymbolsFinder.Instance.GetDerivedTypeSymbolGroupsAsync(
                 memberSymbol,
+                solution,
                 cancellationToken).ConfigureAwait(false);
 
-            // Ensure the user won't be able to see symbol outside the solution for derived symbols.
-            // For example, if user is viewing 'IEnumerable interface' from metadata, we don't want to tell
-            // the user all the derived types under System.Collections
-            var derivedSymbols = allDerivedSymbols.WhereAsArray(symbol => symbol.Locations.Any(static l => l.IsInSource));
-
-            if (baseSymbols.Any() || derivedSymbols.Any())
+            if (memberSymbol.TypeKind == TypeKind.Interface)
             {
-                if (memberSymbol.TypeKind == TypeKind.Interface)
+                if (baseInterfaceSymbolGroups.Any() || derivedTypeSymbolGroups.Any())
                 {
+                    // Interface can only have base interfaces.
+                    Debug.Assert(baseTypeSymbolGroups.IsEmpty);
```

Comment:
this assert is ok.  but it can be worrying as perhaps teh compiler might have other symbols in error conditions.  but fine to keep for now as it's only a debug assert.